apps/hash-graph/lib/graph/src/store/fetcher.rs

```diff
@@ -105,11 +110,249 @@ where
+                references.push(reference);
+            }
+        }
+
+        Ok(references)
+    }
+
+    async fn fetch_external_ontology_types(
+        &self,
+        ontology_type_references: impl IntoIterator<Item = (OntologyTypeReference<'_>, UpdatedById)>,
+    ) -> Result<
+        (
+            Vec<(DataType, OntologyElementMetadata)>,
+            Vec<(PropertyType, OntologyElementMetadata)>,
+            Vec<(EntityType, OntologyElementMetadata)>,
+        ),
+        StoreError,
+    > {
+        let fetcher_client = self.fetcher_client().await?;
+
+        let mut data_types = Vec::new();
+        let mut property_types = Vec::new();
+        let mut entity_types = Vec::new();
+
+        for (reference, fetched_by) in ontology_type_references {
+            let provenance_metadata = ProvenanceMetadata::new(fetched_by);
+            let fetched_ontology_types = fetcher_client
+                .fetch_ontology_type_exhaustive(context::current(), reference.uri().clone())
+                .await
+                .into_report()
+                .change_context(StoreError)?
+                .into_report()
+                .change_context(StoreError)?;
+
+            for fetched_ontology_type in fetched_ontology_types.results {
+                match fetched_ontology_type.ontology_type {
+                    OntologyType::DataType(data_type) => {
+                        let data_type = DataType::try_from(data_type)
+                            .into_report()
+                            .change_context(StoreError)?;
+                        let data_type_reference = DataTypeReference::new(data_type.id().clone());
+                        if !self
+                            .contains_ontology_type(OntologyTypeReference::DataTypeReference(
+                                &data_type_reference,
+                            ))
+                            .await?
+                        {
+                            let metadata = ExternalOntologyElementMetadata::new(
+                                data_type.id().into(),
+                                provenance_metadata,
+                                fetched_ontology_type.fetched_at,
+                            );
+
+                            data_types
+                                .push((data_type, OntologyElementMetadata::External(metadata)));
+                        }
+                    }
+                    OntologyType::PropertyType(property_type) => {
+                        let property_type = PropertyType::try_from(property_type)
+                            .into_report()
+                            .change_context(StoreError)?;
+                        let property_type_reference =
+                            PropertyTypeReference::new(property_type.id().clone());
+                        if !self
+                            .contains_ontology_type(OntologyTypeReference::PropertyTypeReference(
+                                &property_type_reference,
```

Comment:
I wonder if we can run into strange race conditions with there being no outer transaction. Could two tasks be trying to traverse/insert the same ontology types and one would just seemingly crash when trying to insert an entity? If so, would that mean one user would see an error that's actually wrong while another user would see a successful creation??

Reply:
This is a good point and it's likely that you're right. I will do some tests on this and [created a task](https://app.asana.com/0/0/1203958456055107/f) (_internal_).